File: mobile-app/src/domain/models/FileObject.ts
/**
 * Domain Model: Represents a file with all metadata
 * Pure domain model - independent of storage implementation
 */
export interface FileObject {
  id?: string;
  originalName: string;
  contentType: string;
  size: number;
  bucket?: string;
  objectKey?: string;
  publicUrl?: string;
  fileCategory: string;  // 'covers' | 'epubs' | 'pdfs' | 'avatars'
  createdAt?: number;
  updatedAt?: number;
  status?: 'UPLOADING' | 'UPLOADED' | 'DELETED' | 'FAILED';
  ownerId?: string;
}

/**
 * FileUploadRequest - payload for upload API
 *
 * NOTE: In React Native, files are NOT Web File/Blob objects.
 * Use the RNFileDescriptor format from FileStorageApiAdapter.
 * This interface is kept for legacy compatibility only.
 *
 * @deprecated Use RNFileDescriptor from FileStorageApiAdapter directly.
 */
export interface FileUploadRequest {
  /** React Native file URI (content:// or file://) or Web File/Blob */
  file: { uri: string; name: string; type: string } | File | Blob;
  fileCategory: string;
  onProgress?: (percent: number) => void;
}

/**
 * FileUploadResponse - response from server after upload
 */
export interface FileUploadResponse {
  originalFileName: string;
  url: string;
  fileId: string;
  size: number;
  contentType: string;
  uploadedAt: number;
}

/**
 * FileDownloadRequest - parameters for download
 */
export interface FileDownloadRequest {
  fileId: string;
  fileName?: string;
}

File: file-service/src/main/java/com/chiennc/file/domain/models/FileObject.java
package com.chiennc.file.domain.models;

import lombok.*;
import lombok.experimental.FieldDefaults;

/**
 * Domain Model: FileObject represents a file stored in object storage.
 * This is a pure domain concept, independent of implementation details (local disk, S3, MinIO, etc.)
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
public class FileObject {
    
    // Identity
    String id;
    
    // File metadata
    String originalName;
    String contentType;
    long size;
    String md5Checksum;
    
    // Storage location (logical, not physical path)
    String bucket;
    String objectKey;
    
    // Ownership and audit
    String ownerId;
    String fileCategory;  // covers, epubs, pdfs, avatars, etc.
    
    // Access information
    String publicUrl;  // Public accessible URL
    
    // Timestamps
    long createdAt;
    long updatedAt;
    
    // Status
    FileStatus status;
    
    public enum FileStatus {
        UPLOADING,
        UPLOADED,
        DELETED,
        FAILED
    }
}
